tpm2/tpm2.go

```diff
@@ -0,0 +1,1247 @@
+	cmd, err := encodeActivateCredential(activeHandle, keyHandle, activePassword, protectorPassword, credBlob, secret)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := runCommand(rw, tagSessions, cmdActivateCredential, tpmutil.RawBytes(cmd))
+	if err != nil {
+		return nil, err
+	}
+	cred, err := decodeActivateCredential(resp)
+	if err != nil {
+		return nil, fmt.Errorf("decoding ActivateCredential response: %v", err)
+	}
+	return cred, nil
+}
+
+func encodeMakeCredential(protectorHandle Handle, credential, activeName []byte) ([]byte, error) {
+	b1, err := tpmutil.Pack(protectorHandle)
+	if err != nil {
+		return nil, err
+	}
+	b2, err := tpmutil.Pack(credential, activeName)
+	if err != nil {
+		return nil, err
+	}
+	return append(b1, b2...), nil
+}
+
+func decodeMakeCredential(in []byte) ([]byte, []byte, error) {
+	var credBlob []byte
+	var encryptedSecret []byte
+
+	err := tpmutil.Unpack(in, &credBlob, &encryptedSecret)
+	if err != nil {
+		return nil, nil, err
+	}
+	return credBlob, encryptedSecret, nil
+}
+
+// MakeCredential creates an encrypted credential for use in MakeCredential.
+// Returns encrypted credential and wrapped secret used to encrypt it.
+func MakeCredential(rw io.ReadWriter, protectorHandle Handle, credential, activeName []byte) ([]byte, []byte, error) {
+	cmd, err := encodeMakeCredential(protectorHandle, credential, activeName)
+	if err != nil {
+		return nil, nil, err
+	}
+	resp, err := runCommand(rw, tagNoSessions, cmdMakeCredential, tpmutil.RawBytes(cmd))
+	if err != nil {
+		return nil, nil, err
+	}
+	credBlob, encryptedSecret, err := decodeMakeCredential(resp)
+	if err != nil {
+		return nil, nil, fmt.Errorf("decoding MakeCredential response: %v", err)
+	}
+	return credBlob, encryptedSecret, nil
+}
+
+func encodeEvictControl(owner Handle, tmpHandle, persistantHandle Handle) ([]byte, error) {
+	b1, err := tpmutil.Pack(owner)
+	if err != nil {
+		return nil, err
+	}
+	b2, err := tpmutil.Pack(tmpHandle)
+	if err != nil {
+		return nil, err
+	}
+	b3, err := tpmutil.Pack([]byte(nil))
+	if err != nil {
+		return nil, err
+	}
+	b4, err := encodePasswordAuthArea("", HandlePasswordSession)
+	if err != nil {
+		return nil, err
+	}
+	b5, err := tpmutil.Pack(persistantHandle)
+	if err != nil {
+		return nil, err
+	}
+	args := append(b1, b2...)
+	args = append(args, b3...)
+	args = append(args, b4...)
+	args = append(args, b5...)
+	return args, nil
+}
+
+// EvictControl toggles persistence of an object within the TPM.
+func EvictControl(rw io.ReadWriter, owner, tmpHandle, persistantHandle Handle) error {
+	cmd, err := encodeEvictControl(owner, tmpHandle, persistantHandle)
+	if err != nil {
+		return err
+	}
+	_, err = runCommand(rw, tagSessions, cmdEvictControl, tpmutil.RawBytes(cmd))
+	return err
+}
+
+// ContextSave returns an encrypted version of the session, object or sequence
+// context for storage outside of the TPM.
```

Comment:
Please add a little more context to the comment. I think it must be the case that the choice of session, object, or sequence depends on what the Handle refers to. Is that right?

Reply:
Yes, handle points to _something_ that is being saved.